aot autograd: perform input mutations in no_grad()

bdhirsh opened the pull request:
This is a (potential) fix for https://github.com/pytorch/pytorch/issues/90844.

What's kinda unfortunate is that `aot_module()` won't error out anymore when you try to mutate leaves that require grad. I added a test that confirms that dynamo will still error out.

I think we agreed that `aot_module()` is a private API, so maybe this separation is acceptable? Open to feedback. I think getting the error to show up in aot autograd properly will take some more work.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90850



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

bdhirsh committed:
aot autograd: perform input mutations in no_grad()

[ghstack-poisoned]

ezyang commented:
Aot_module is public API but I also don't want to fix it right now lol

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1803,7 +1803,8 @@ def compiled_function(*args):
                         )
-                    original_inpt.copy_(updated_inpt)
+                    with torch.no_grad():
+                        original_inpt.copy_(updated_inpt)
```
It's not obvious to me why disabling autograd here is correct. The old code you could backprop through subsequent uses of input; with the new code you can't, right?

bdhirsh replied:
oh you're totally right. I think I should do something like:
```
if original_inpt.is_leaf:
    ctx = torch.no_grad()
else:
    ctx = nullcontext()

with ctx:
    original_inpt.copy_(updated_inpt)
```

bdhirsh replied:
and CI right now is correctly failing, for the reason you described

bdhirsh replied:
(updated, I see tests passing locally)

ezyang replied:
Mmmf, I am willing to accept this to unblock but it still seems fishy. If no grad was still set on a subsequent call, then this shouldn't have been necessary. But if the no grad context changed, we in fact need to guard fail and recompile, no?

bdhirsh replied:
hmmm.... I tried to repro the example that you gave, and it looks to me like dynamo is guarding properly. Unless I'm misunderstanding?
```

# I confirmed that this repro compiles twice - indicating the dynamo recompiles on the `torch.no_grad()` context change?
@torch.compile(backend="aot_eager")
def f(a):
    a.clone().add_(1)
    return a

inp = [torch.ones(3, requires_grad=True)]
with torch.no_grad():
    out = f(*inp)
out = f(*inp)
```

```
# I confirmed that this compiles once, and then erorrs out on the second call of f(), where we are no longer in no_grad mode.
@torch.compile(backend="aot_eager")
def f(a):
    a.add_(1)
    return a

inp = [torch.ones(3, requires_grad=True)]
with torch.no_grad():
    out = f(*inp)
out = f(*inp)
```

ezyang replied:
So then why do we need to patch it here 🤔

bdhirsh replied:
the case that this PR fixes is when the no_grad() is **inside** the function we're compiling:

```
@torch.compile(backend='aot_eager')
def f(x):
    with torch.no_grad():
        x.add_(1)

x = torch.zeros(2, requires_grad=True)
# Fails with "RuntimeError: a leaf Variable that requires grad is being used in an in-place operation"
out = f(x)
```

ezyang replied:
Offline discussion: we need to no grad even if we mutate a non-leaf tensor, so Brian is going to try to track this directyl.

ezyang replied:
I think the other edge case I'm worried about is something like this:

```
x.add_(y)
with no_grad():
    x.add_(y2)
x.add_(y3)
```

It's not clear to me that we can actually express this as a single `copy_` outside of the custom autograd function anymore...

ezyang replied:
@albanD convinced me this is fine, you just have to make sure you have a funky grad for x which only has the backwards contribution from y and y3 but not y2.

ezyang replied:
So for this particular example, you want the backward to be

```
grad_x = grad_y + grad_y3
```

as the y2 addition being untracked due to the no grad. So all you need to do is make sure *this* is what goes in your backward graph, and then it's fine to set x's grad_fn for this.

bdhirsh committed:
Update on "aot autograd: perform input mutations in no_grad()"

This is a (potential) fix for https://github.com/pytorch/pytorch/issues/90844.

What's kinda unfortunate is that `aot_module()` won't error out anymore when you try to mutate leaves that require grad. I added a test that confirms that dynamo will still error out.

I think we agreed that `aot_module()` is a private API, so maybe this separation is acceptable? Open to feedback. I think getting the error to show up in aot autograd properly will take some more work.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
